clients/imodels-client-authoring/src/operations/changeset/ChangesetOperations.ts

```diff
@@ -198,4 +240,38 @@ export class ChangesetOperations<TOptions extends OperationOptions> extends Mana
   private createFileName(changesetId: string): string {
     return `${changesetId}.cs`;
   }
+
+  private async provideDownloadCallbacks(params: DownloadChangesetListParams): Promise<[DownloadCallback, DownloadFailedCallback] | undefined> {
+    if (!params.progressCallback)
+      return;
+
+    let totalSize = 0;
+    let totalDownloaded = 0;
+
+    for await (const changesetPage of this.getMinimalList(params).byPage()){
+      for (const changeset of changesetPage){
+        totalSize += changeset.fileSize;
+
+        const filePath = path.join(params.targetDirectoryPath, this.createFileName(changeset.id));
+        if (await this.isChangesetAlreadyDownloaded(filePath, changeset.fileSize))
+          totalDownloaded += changeset.fileSize;
+      }
+    }
+
+    const progressCallback: DownloadCallback = (downloaded) => {
+      totalDownloaded += downloaded;
+      params.progressCallback?.(totalDownloaded, totalSize);
+    };
+    const failureCallback: DownloadCallback = (downloadedBeforeFailure) => totalSize += downloadedBeforeFailure;
```

Comment:
> I am not a fan of code comments but I think this line here would benefit from a short explanation on why this additional handling is needed (to not report more bytes downloaded than there are total in case of failure on first try). It is clear to us now since we discussed this a bit but I assume all this handling could look excessive for someone looking at this for the first time.

Added.

> Also maybe name `on1stAttemptFailed` would be more clear?

Renamed to `downloadFailedCallback` instead. Name `on1stAttemptFailed` would be a bit misleading, since in `provideDownloadCallbacks` context `downloadFailedCallback` theoretically can handle _2nd_, _3rd_ and _nth_ failure.